File: python/anagram/anagram.py
class Anagram(object):

    def __init__(self, word):
        self.word = word
        self._lower = word.lower()
        self._normalized = self.__normalize(self._lower)

    @staticmethod
    def __normalize(word):
        """Return the letters of a word sorted and lowercased."""
        return sorted(word.lower())

    def is_anagram(self, candidate):
        """Check if candidate is an anagram of self.word."""
        return ((self.__normalize(candidate) == self._normalized) and
                (candidate.lower() != self._lower))

    def match(self, candidates):
        """Return only those candidates that are anagrams of self.word."""
        return [c for c in candidates if self.is_anagram(c)]

File: haskell/anagram/Anagram.hs
module Anagram (anagramsFor) where


import Data.Char (toLower)
import Data.List (sort)


-- | 'anagramsFor', applied to a word and a list of candidate strings,
-- returns the list of those candidates that are anagrams of the word.
anagramsFor :: String -> [String] -> [String]
anagramsFor x ys = filter (isAnagramOfX . lowerCase) ys
  where isAnagramOfX yLower = yLower /= xLower && sort yLower == xLowerSorted
        xLower              = lowerCase x
        xLowerSorted        = sort xLower


lowerCase :: String -> String
lowerCase = map toLower
